File: app/src/main/java/com/panoramagl/sample/MainActivity.java
package com.panoramagl.sample;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import com.panoramagl.PLConstants;
import com.panoramagl.PLICamera;
import com.panoramagl.PLImage;
import com.panoramagl.PLManager;
import com.panoramagl.PLSphericalPanorama;
import com.panoramagl.hotspots.ActionPLHotspot;
import com.panoramagl.hotspots.HotSpotListener;
import com.panoramagl.utils.PLUtils;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity implements HotSpotListener {

    private PLManager plManager;
    private int currentIndex = -1;
    private int[] resourceIds = new int[]{R.raw.sighisoara_sphere, R.raw.sighisoara_sphere_2};

    private View.OnClickListener buttonClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            switch (v.getId()) {
                case R.id.button_1:
                    changePanorama(0);
                    break;
                case R.id.button_2:
                    changePanorama(1);
                    break;
                default:
                    break;
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        plManager = new PLManager(this);
        plManager.setContentView((ViewGroup) findViewById(R.id.content_view));
        plManager.onCreate();

        plManager.setAccelerometerEnabled(false);
        plManager.setInertiaEnabled(false);
        plManager.setZoomEnabled(false);

        changePanorama(0);

        Button button1 = findViewById(R.id.button_1);
        Button button2 = findViewById(R.id.button_2);
        button1.setOnClickListener(buttonClickListener);
        button2.setOnClickListener(buttonClickListener);
    }

    @Override
    protected void onResume() {
        super.onResume();
        plManager.onResume();
    }

    @Override
    protected void onPause() {
        plManager.onPause();
        super.onPause();
    }

    @Override
    protected void onDestroy() {
        plManager.onDestroy();
        super.onDestroy();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return plManager.onTouchEvent(event);
    }

    private void changePanorama(int index) {
        if (currentIndex == index) return;

        Bitmap image3D = PLUtils.getBitmap(this, resourceIds[index]);

        PLSphericalPanorama panorama = new PLSphericalPanorama();
        panorama.setImage(new PLImage(image3D, false));
        float pitch = 5f;
        float yaw = 0f;
        float zoomFactor = 0.7f;

        if (currentIndex != -1) {
            PLICamera camera = plManager.getPanorama().getCamera();
            pitch = camera.getPitch();
            yaw = camera.getYaw();
            zoomFactor = camera.getZoomFactor();
        }

        panorama.removeAllHotspots();
        long hotSpotId = 100;


        float normalizedX = 500f / image3D.getWidth();
        float normalizedY = 700f / image3D.getHeight();
        ActionPLHotspot plHotspot1 = new ActionPLHotspot(this, hotSpotId, new PLImage(BitmapFactory.decodeResource(getResources(), R.raw.hotspot)), 0, 0, PLConstants.kDefaultHotspotSize, PLConstants.kDefaultHotspotSize);
        plHotspot1.setPosition(normalizedX, normalizedY);

        ActionPLHotspot plHotspot2 = new ActionPLHotspot(this, hotSpotId + 1, new PLImage(BitmapFactory.decodeResource(getResources(), R.raw.hotspot)), 20, 50, PLConstants.kDefaultHotspotSize, PLConstants.kDefaultHotspotSize);
        panorama.addHotspot(plHotspot1);
        panorama.addHotspot(plHotspot2);

        panorama.getCamera().lookAtAndZoomFactor(pitch, yaw, zoomFactor, false);
        plManager.setPanorama(panorama);
        currentIndex = index;
    }

    @Override
    public void onClick(final long identifier) {
        runOnUiThread(
                new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(MainActivity.this, "HotSpotClicked! Id is-> " + identifier, Toast.LENGTH_SHORT).show();
                    }
                }
        );
    }


}

File: app/src/main/java/com/panoramagl/sample/MainActivity.kt
package com.panoramagl.sample

import android.graphics.BitmapFactory
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.MotionEvent
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.panoramagl.*
import com.panoramagl.hotspots.ActionPLHotspot
import com.panoramagl.hotspots.HotSpotListener
import com.panoramagl.sample.databinding.ActivityMainBinding
import com.panoramagl.utils.PLUtils
import timber.log.Timber

class MainActivity : AppCompatActivity(), HotSpotListener {

    private lateinit var binding: ActivityMainBinding

    private lateinit var plManager: PLManager
    private var currentIndex = -1
    private val resourceIds = intArrayOf(R.raw.sighisoara_sphere, R.raw.sighisoara_sphere_2)

    private val useAcceleratedTouchScrolling = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        plManager = PLManager(this).apply {
            setContentView(binding.contentView)

            onCreate()
            isAccelerometerEnabled = false
            isInertiaEnabled = false
            isZoomEnabled = false
            isAcceleratedTouchScrollingEnabled = useAcceleratedTouchScrolling
            // to see a black screen
            isScrollingEnabled = true
            isInertiaEnabled = true
        }
        changePanorama(0)
        binding.button1.setOnClickListener { changePanorama(0) }
        binding.button2.setOnClickListener { changePanorama(1) }

        binding.contentView.setOnLongClickListener {
            val x = it.x
            val y = it.y
            val screenWidth = resources.displayMetrics.widthPixels
            val screenHeight = resources.displayMetrics.heightPixels
            val normalizedX = x / screenWidth
            val normalizedY = y / screenHeight

            Timber.d("$normalizedX/$normalizedY $screenWidth/$screenHeight")
            addHotspotAt(normalizedX, normalizedY, screenWidth, screenHeight)
            false
        }
    }

    private fun addHotspotAt(normalizedX: Float, normalizedY: Float, screenWidth: Int, screenHeight: Int) {
        // These values would ideally come from your panorama viewer or camera settings
        val currentYaw = plManager.panorama.camera.yaw
        val currentPitch = plManager.panorama.camera.pitch
        val fov = plManager.panorama.camera.fov  // Current FOV considering zoom
        val aspectRatio = screenWidth.toFloat() / screenHeight

        // Convert normalized screen coordinates to changes in yaw and pitch
        val yawChange = (normalizedX - 0.5f) * fov
        val pitchChange = (0.5f - normalizedY) * (fov / aspectRatio)

        // Calculate new yaw and pitch for the hotspot
        val hotspotYaw = (currentYaw + yawChange) % 360
        val hotspotPitch = currentPitch + pitchChange
        // Ensure pitch stays within bounds [-90, 90]
        val clampedPitch = hotspotPitch.coerceIn(-90f, 90f)

        // Now, create your hotspot at this new yaw and pitch
        createHotspot(hotspotPitch = clampedPitch, hotspotYaw = hotspotYaw)
    }

    private fun createHotspot(hotspotPitch: Float, hotspotYaw: Float) {
        val hotspot = ActionPLHotspot(
            this,
            System.currentTimeMillis(),
            PLImage(BitmapFactory.decodeResource(resources, R.raw.hotspot)),
            hotspotPitch,
            hotspotYaw,
            PLConstants.kDefaultHotspotSize,
            PLConstants.kDefaultHotspotSize
        )
        plManager.panorama.addHotspot(hotspot)
    }

    override fun onResume() {
        super.onResume()
        plManager.onResume()
        Handler(Looper.getMainLooper()).postDelayed({
            binding.textVersion.text = PLRenderer.versionGL
            binding.textRenderer.text = PLRenderer.rendererGL
        }, 1000)
    }

    override fun onPause() {
        plManager.onPause()
        super.onPause()
    }

    override fun onDestroy() {
        plManager.onDestroy()
        super.onDestroy()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return plManager.onTouchEvent(event)
    }

    private fun changePanorama(index: Int) {
        if (currentIndex == index)
            return
        val image3D = PLUtils.getBitmap(this, resourceIds[index])
        val panorama = PLSphericalPanorama()
        panorama.setImage(PLImage(image3D, false))
        var pitch = 5f
        var yaw = 0f
        var zoomFactor = 0.7f
        if (currentIndex != -1) {
            plManager.panorama.camera?.apply {
                pitch = this.pitch
                yaw = this.yaw
                zoomFactor = this.zoomFactor
            }
        }
        panorama.removeAllHotspots()
        val hotSpotId: Long = 100
        val normalizedX = 500f / image3D.width
        val normalizedY = 700f / image3D.height
        val plHotspot1 = ActionPLHotspot(
            this,
            hotSpotId,
            PLImage(BitmapFactory.decodeResource(resources, R.raw.hotspot)),
            0f,
            0f,
            PLConstants.kDefaultHotspotSize,
            PLConstants.kDefaultHotspotSize
        )
        plHotspot1.setPosition(normalizedX, normalizedY)
        val plHotspot2 = ActionPLHotspot(
            this,
            hotSpotId + 1,
            PLImage(BitmapFactory.decodeResource(resources, R.raw.hotspot)),
            20f,
            50f,
            PLConstants.kDefaultHotspotSize,
            PLConstants.kDefaultHotspotSize
        )
        panorama.addHotspot(plHotspot1)
        panorama.addHotspot(plHotspot2)
        panorama.camera.lookAtAndZoomFactor(pitch, yaw, zoomFactor, false)
        if (!useAcceleratedTouchScrolling) {
            // If not using the accelerated scrolling, increasing the camera's rotation sensitivity will allow the
            // image to pan faster with finger movement. 180f gives about a ~1:1 move sensitivity.
            // Higher will move the map faster
            // Range 1-270
            panorama.camera.rotationSensitivity = 270f
        }
        plManager.panorama = panorama
        currentIndex = index
        plManager.startSensorialRotation()
    }

    override fun onHotspotClick(identifier: Long) {
        runOnUiThread { Toast.makeText(this@MainActivity, "HotSpotClicked! Id is-> $identifier", Toast.LENGTH_SHORT).show() }
    }
}
